src/object/obj_tx.c

```diff
@@ -0,0 +1,1047 @@
+
+		/* If the pool map is refreshed, then it is possible that
+		 * the data from former fetch/list/query become stale. On
+		 * the other hand, even if related data is still valid,
+		 * but related read timestamp (used for MVCC) may be left
+		 * on the server that has been evicted from the cluster.
+		 * So we have to restart the transaction under such case.
+		 */
+		dc_tx_restart_internal(tx, 0);
+		rc = -DER_AGAIN;
+	}
+
+out:
+	if (rc != 0 || ptx == NULL)
+		dc_tx_decref(tx);
+	else
+		*ptx = tx;
+
+	return rc;
+}
+
+int
+dc_tx_check_pmv(daos_handle_t th)
+{
+	return dc_tx_check_pmv_internal(th, NULL);
+}
+
+static int
+dc_tx_check(daos_handle_t th, bool check_write, struct dc_tx **ptx)
+{
+	struct dc_tx	*tx = NULL;
+	int		 rc;
+
+	rc = dc_tx_check_pmv_internal(th, &tx);
+	if (rc != 0)
+		return rc;
+
+	if (daos_handle_is_inval(th))
+		return -DER_INVAL;
+
+	if (tx->tx_status != TX_OPEN) {
+		D_ERROR("TX is not valid for update.\n");
+		D_GOTO(out, rc = -DER_NO_PERM);
+	}
+
+	if (check_write && tx->tx_flags & DTF_RDONLY) {
+		D_ERROR("TX is READ ONLY\n");
+		D_GOTO(out, rc = -DER_NO_PERM);
+	}
+
+out:
+	if (rc != 0)
+		dc_tx_decref(tx);
+	else
+		*ptx = tx;
+
+	return rc;
+}
+
+int
+dc_tx_hdl2epoch_and_pmv(daos_handle_t th, daos_epoch_t *epoch, uint32_t *pm_ver)
+{
+	struct dc_tx	*tx = NULL;
+	int		 rc;
+
+	rc = dc_tx_check(th, false, &tx);
+	if (rc == 0) {
+		if (tx->tx_pm_ver == 0)
+			tx->tx_pm_ver = dc_pool_get_version_lock(tx->tx_pool);
+
+		*pm_ver = tx->tx_pm_ver;
+		*epoch = tx->tx_epoch;
+		dc_tx_decref(tx);
+	}
+
+	return rc;
+}
+
+int
+dc_tx_hdl2epoch(daos_handle_t th, daos_epoch_t *epoch)
+{
+	struct dc_tx	*tx;
+	int		 rc = 0;
+
+	if (daos_handle_is_inval(th))
+		return -DER_INVAL;
+
+	tx = dc_tx_hdl2ptr(th);
+	if (tx == NULL)
+		return -DER_NO_HDL;
+
+	/**
+	 * XXX: If the TX has never talked with any server, its epoch will
+	 *	be zero, dc_tx_hdl2epoch() returns -DER_UNINIT to indicate
+	 *	that. The caller can re-call hdl2epoch after some fetch or
+	 *	TX commit.
+	 */
+	if (tx->tx_epoch != 0)
+		*epoch = tx->tx_epoch;
+	else
+		rc = -DER_UNINIT;
+	dc_tx_decref(tx);
+
+	return rc;
+}
+
+int
+dc_tx_open(tse_task_t *task)
+{
+	daos_tx_open_t	*args;
+	struct dc_tx	*tx = NULL;
+	int		 rc;
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL,
+		  "Task Argument OPC does not match DC OPC (open)\n");
+
+	rc = dc_tx_alloc(args->coh, 0, 0, &tx);
+	if (rc == 0)
+		*args->th = dc_tx_ptr2hdl(tx);
+
+	tse_task_complete(task, rc);
+
+	return rc;
+}
+
+struct tx_commit_cb_args {
+	struct dc_tx	*tcca_tx;
+	crt_rpc_t	*tcca_req;
+};
+
+static int
+dc_tx_commit_cb(tse_task_t *task, void *data)
+{
+	struct tx_commit_cb_args *tcca = (struct tx_commit_cb_args *)data;
+	struct dc_tx		 *tx = tcca->tcca_tx;
+	crt_rpc_t		 *req = tcca->tcca_req;
+	int			  rc = task->dt_result;
+
+	if (rc == 0) {
+		tx->tx_status = TX_COMMITTED;
+		goto out;
+	}
+
+	/* Need to restart the TX with newer epoch. */
+	if (rc == -DER_AGAIN) {
```

Comment:
Do we automatically restart the user transaction? or should the user restart it?

Reply:
Since we do not export restart() API, then DTX client side logic will auto restart the DTX if server indicates that via -DER_AGAIN. The restart logic will cleanup the cached modifications and assign new epoch.

Reply:
If we want user to restart the TX, then two options:
1) we export TX restart() API, but seems others (Liwei, Johann) do not like such idea.
2) close the transaction firstly, then open a new TX.